testresult: correctly apply verbose word markup and avoid crash

pbrezina opened the pull request:
The following snippet would have resulted in crash on multiple places since
`_get_verbose_word` expects only string, not a tuple.

```python
    @pytest.hookimpl(tryfirst=True)
    def pytest_report_teststatus(report: pytest.CollectReport | pytest.TestReport, config: pytest.Config):
        if report.when == "call":
            return ("error", "A",  ("AVC", {"bold": True, "red": True}))

        return None
```

```python-traceback
Traceback (most recent call last):
  File "/home/pbrezina/workspace/sssd/.venv/bin/pytest", line 8, in <module>
    sys.exit(console_main())
             ^^^^^^^^^^^^^^
  File "/home/pbrezina/workspace/pytest/src/_pytest/config/__init__.py", line 207, in console_main
    code = main()
           ^^^^^^
  File "/home/pbrezina/workspace/pytest/src/_pytest/config/__init__.py", line 179, in main
    ret: Union[ExitCode, int] = config.hook.pytest_cmdline_main(
                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/pbrezina/workspace/sssd/.venv/lib64/python3.11/site-packages/pluggy/_hooks.py", line 513, in __call__
    return self._hookexec(self.name, self._hookimpls.copy(), kwargs, firstresult)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/pbrezina/workspace/sssd/.venv/lib64/python3.11/site-packages/pluggy/_manager.py", line 120, in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/pbrezina/workspace/sssd/.venv/lib64/python3.11/site-packages/pluggy/_callers.py", line 139, in _multicall
    raise exception.with_traceback(exception.__traceback__)
  File "/home/pbrezina/workspace/sssd/.venv/lib64/python3.11/site-packages/pluggy/_callers.py", line 103, in _multicall
    res = hook_impl.function(*args)
          ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/pbrezina/workspace/pytest/src/_pytest/main.py", line 333, in pytest_cmdline_main
    return wrap_session(config, _main)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/pbrezina/workspace/pytest/src/_pytest/main.py", line 321, in wrap_session
    config.hook.pytest_sessionfinish(
  File "/home/pbrezina/workspace/sssd/.venv/lib64/python3.11/site-packages/pluggy/_hooks.py", line 513, in __call__
    return self._hookexec(self.name, self._hookimpls.copy(), kwargs, firstresult)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/pbrezina/workspace/sssd/.venv/lib64/python3.11/site-packages/pluggy/_manager.py", line 120, in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/pbrezina/workspace/sssd/.venv/lib64/python3.11/site-packages/pluggy/_callers.py", line 139, in _multicall
    raise exception.with_traceback(exception.__traceback__)
  File "/home/pbrezina/workspace/sssd/.venv/lib64/python3.11/site-packages/pluggy/_callers.py", line 122, in _multicall
    teardown.throw(exception)  # type: ignore[union-attr]
    ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/pbrezina/workspace/pytest/src/_pytest/logging.py", line 872, in pytest_sessionfinish
    return (yield)
            ^^^^^
  File "/home/pbrezina/workspace/sssd/.venv/lib64/python3.11/site-packages/pluggy/_callers.py", line 124, in _multicall
    teardown.send(result)  # type: ignore[union-attr]
    ^^^^^^^^^^^^^^^^^^^^^
  File "/home/pbrezina/workspace/pytest/src/_pytest/terminal.py", line 899, in pytest_sessionfinish
    self.config.hook.pytest_terminal_summary(
  File "/home/pbrezina/workspace/sssd/.venv/lib64/python3.11/site-packages/pluggy/_hooks.py", line 513, in __call__
    return self._hookexec(self.name, self._hookimpls.copy(), kwargs, firstresult)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/pbrezina/workspace/sssd/.venv/lib64/python3.11/site-packages/pluggy/_manager.py", line 120, in _hookexec
    return self._inner_hookexec(hook_name, methods, kwargs, firstresult)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/pbrezina/workspace/sssd/.venv/lib64/python3.11/site-packages/pluggy/_callers.py", line 139, in _multicall
    raise exception.with_traceback(exception.__traceback__)
  File "/home/pbrezina/workspace/sssd/.venv/lib64/python3.11/site-packages/pluggy/_callers.py", line 124, in _multicall
    teardown.send(result)  # type: ignore[union-attr]
    ^^^^^^^^^^^^^^^^^^^^^
  File "/home/pbrezina/workspace/pytest/src/_pytest/terminal.py", line 923, in pytest_terminal_summary
    self.short_test_summary()
  File "/home/pbrezina/workspace/pytest/src/_pytest/terminal.py", line 1272, in short_test_summary
    action(lines)
  File "/home/pbrezina/workspace/pytest/src/_pytest/terminal.py", line 1205, in show_simple
    line = _get_line_with_reprcrash_message(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/pbrezina/workspace/pytest/src/_pytest/terminal.py", line 1429, in _get_line_with_reprcrash_message
    word = tw.markup(verbose_word, **word_markup)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/pbrezina/workspace/pytest/src/_pytest/_io/terminalwriter.py", line 114, in markup
    text = "".join(f"\x1b[{cod}m" for cod in esc) + text + "\x1b[0m"
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~
TypeError: can only concatenate str (not "tuple") to str
```



<!--
Thanks for submitting a PR, your contribution is really appreciated!

Here is a quick checklist that should be present in PRs.

- [ ] Include documentation when adding new features.
- [ ] Include new tests or update existing tests when applicable.
- [X] Allow maintainers to push and squash when merging my commits. Please uncheck this if you prefer to squash the commits yourself.

If this change fixes an issue, please:

- [ ] Add text like ``closes #XYZW`` to the PR description and/or commits (where ``XYZW`` is the issue number). See the [github docs](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) for more information.

Unless your change is trivial or a small documentation fix (e.g., a typo or reword of a small section) please:

- [ ] Create a new changelog file in the `changelog` folder, with a name like `<ISSUE NUMBER>.<TYPE>.rst`. See [changelog/README.rst](https://github.com/pytest-dev/pytest/blob/main/changelog/README.rst) for details.

  Write sentences in the **past or present tense**, examples:

  * *Improved verbose diff output with sequences.*
  * *Terminal summary statistics now use multiple colors.*

  Also make sure to end the sentence with a `.`.

- [ ] Add yourself to `AUTHORS` in alphabetical order.
-->

The-Compiler reviewed (changes requested):
Thank you! This looks reasonable, but will need a test case to avoid similar regressions in the future.

pbrezina commented:
> Thank you! This looks reasonable, but will need a test case to avoid similar regressions in the future.

Please, could you contribute a test case by pushing to this pull request? I guess I could base something on `test_line_with_reprcrash` but this will only trigger one of the cases.

The-Compiler commented:
I'm afraid I got my hands full with stuff we're currently working on at the development sprint already.

It might be easier to write an integration test using `pytester` for this. Actually I'm confused: We already have `test_report_teststatus_explicit_markup` which is very close to what you showed in your original post, but that passes! So the first step would probably be to find out why that passes just fine. Maybe it just needs an additional check for the exit status or something?

I tried minimizing your example to:

```python
def pytest_report_teststatus(report):
    return ("error", "A",  ("AVC", {"bold": True, "red": True}))
```

and I can still reproduce the failure, yet the test seems to pass just fine, despite doing almost the same:

```python
        pytester.makeconftest(
            """
            def pytest_report_teststatus(report):
                return 'foo', 'F', ('FOO', {'red': True})
        """
        )
```

webknjaz commented:
@pbrezina please also compose a change log note/fragment, while you're on it. It's important to explain how the behavior changed to the end-users.

pbrezina commented:
@The-Compiler it fails only with error or failed category, not with foo. And if you set error in the test, the line matcher still works since the traceback is in stderr. I fixed the test and added changelog.

webknjaz commented on src/_pytest/reports.py:
```diff
@@ -196,11 +196,17 @@ def head_line(self) -> Optional[str]:
         return None
 
-    def _get_verbose_word(self, config: Config):
+    def _get_verbose_word(
```
As the method does not return a single word anymore, it should probably be renamed to reflect what it does, really. Perhaps, smth like this:
```suggestion
    def _get_verbose_word_with_markup(
```

pbrezina replied:
Done.

webknjaz commented on testing/test_terminal.py:
```diff
@@ -327,16 +327,17 @@ def test_rewrite(self, pytester: Pytester, monkeypatch) -> None:
+            f"""
             def pytest_report_teststatus(report):
-                return 'foo', 'F', ('FOO', {'red': True})
+                return '{category}', 'F', ('FOO', {{'red': True}})
```
It's probably safer to use a string repr for cases when it gets to have a single quote inside for whatever reason:
```suggestion
                return {category !r}, 'F', ('FOO', {{'red': True}})
```

webknjaz commented on src/_pytest/reports.py:
```diff
@@ -196,11 +196,17 @@ def head_line(self) -> Optional[str]:
+    def _get_verbose_word(
+        self, config: Config, default_markup: Mapping[str, bool]
+    ) -> Tuple[str, Mapping[str, bool]]:
         _category, _short, verbose = config.hook.pytest_report_teststatus(
```
*(non-blocking)* `verbose` here might benefit from being renamed to something more verbose. Like `word_and_maybe_markup`.

pbrezina replied:
I used _get_verbose_word_with_markup suggested in different comment since markup is always returned.

webknjaz commented on src/_pytest/reports.py:
```diff
@@ -196,11 +196,17 @@ def head_line(self) -> Optional[str]:
-    def _get_verbose_word(self, config: Config):
+    def _get_verbose_word(
+        self, config: Config, default_markup: Mapping[str, bool]
+    ) -> Tuple[str, Mapping[str, bool]]:
```
The pre-commit run here https://results.pre-commit.ci/run/github/37489525/1719242655.jVEQ62HET5qmvF-sLT8qtA exposes that `Tuple` hasn't been imported. Though, perhaps a lowercase variant would work, since Ronny pushed some typing improvements last week:
```suggestion
    ) -> tuple[str, Mapping[str, bool]]:
```

webknjaz replied:
@pbrezina it used to be imported last week. But Ronny contributed a huge refactoring with future imports that enabled the project to use newer syntax. And dropping the imports was a part of that, as they became unnecessary.

You created your branch just before that happened, so you don't see the changes in your fork. But if you rebase on top of the upstream main branch, you'll see that the imports are no longer there.

pbrezina commented on src/_pytest/reports.py:
```diff
@@ -196,11 +196,17 @@ def head_line(self) -> Optional[str]:
         )
-        return verbose
+
+        if not isinstance(verbose, tuple):
```
Done.

pbrezina commented:
All comments should be addressed now.

webknjaz commented on changelog/12472.bugfix.rst:
```diff
@@ -0,0 +1 @@
+Fixed a crash when returning category ``"error"`` or ``"failed"`` with a custom test status from :hook:`pytest_report_teststatus` hook -- :user:`pbrezina`.
```
FTR here's the change note preview: https://pytest--12472.org.readthedocs.build/en/12472/changelog.html#bug-fixes

webknjaz commented:
All's good on my side, let's wait for Florian and perhaps somebody else to re-review.

nicoddemus commented on src/_pytest/reports.py:
```diff
@@ -190,11 +190,17 @@ def head_line(self) -> str | None:
+        if isinstance(verbose, str):
+            return verbose, default_markup
+
+        return verbose[0], verbose[1]
```
While at it, we should probably also verify that verbose what we expect and raise an error otherwise:

```suggestion
        if isinstance(verbose, Sequence) and len(verbose) == 2:
            word, markup = verbose
            if isinstance(word, str) and isinstance(markup, Mapping):
                return word, markup
        pytest.fail("pytest_report_teststatus() hook (from a plugin) returned an invalid verbose value: {verbose!r}.\nExpected either a string or a tuple of (word, markup)."
```

nicoddemus replied:
(being able to use `match` here would be awesome, but alas only Python 3.10+).

nicoddemus reviewed (approved):
Awesome, thanks!

nicoddemus commented on src/_pytest/reports.py:
```diff
@@ -190,11 +192,26 @@ def head_line(self) -> str | None:
+
+        fail(
+            "pytest_report_teststatus() hook (from a plugin) returned "
+            "an invalid verbose value: {verbose!r}.\nExpected either a string "
```
```suggestion
            f"an invalid verbose value: {verbose!r}.\nExpected either a string "
```

nicoddemus commented on src/_pytest/reports.py:
```diff
@@ -190,11 +192,26 @@ def head_line(self) -> str | None:
+            if isinstance(word, str) and isinstance(markup, Mapping):
+                return word, markup
+
+        fail(
```
```suggestion
        fail(  # pragma: no cover
```

pbrezina commented:
Rebased.

nicoddemus commented:
@webknjaz thanks for merging.

In the future consider squashing when the commits are not atomic/make sense, otherwise we end up with a polluted history:

![image](https://github.com/pytest-dev/pytest/assets/1085180/0fefc73f-51a8-49f7-9201-4a5b4724021b)

nicoddemus commented on src/_pytest/reports.py:
```diff
@@ -190,11 +192,26 @@ def head_line(self) -> str | None:
+                return word, markup
+
+        fail(
+            "pytest_report_teststatus() hook (from a plugin) returned "
```
Yep, I figured that once the formatter fixed it, hehehe.

webknjaz commented:
> In the future consider squashing when the commits are not atomic/make sense, otherwise we end up with a polluted history:

Ah, fair. I usually do this, but this time I forgot to check. I saw that your auto-merge got reset and clicked again w/o noticing that it was in a different mode, not natural merge.
